pkg/webhook/storageclass.go

```diff
@@ -0,0 +1,174 @@
+		return reviewResponse
+	}
+	isUpdate := ar.Request.Operation == v1.Update
+
+	raw := ar.Request.Object.Raw
+	oldRaw := ar.Request.OldObject.Raw
+
+	deserializer := codecs.UniversalDeserializer()
+	switch ar.Request.Resource {
+	case StorageClassV1GVR:
+		sc := &storagev1.StorageClass{}
+		if _, _, err := deserializer.Decode(raw, nil, sc); err != nil {
+			klog.Error(err)
+			return toV1AdmissionResponse(err)
+		}
+		oldSc := &storagev1.StorageClass{}
+		if _, _, err := deserializer.Decode(oldRaw, nil, oldSc); err != nil {
+			klog.Error(err)
+			return toV1AdmissionResponse(err)
+		}
+		return decideV1StorageClass(sc, oldSc, isUpdate)
+	default:
+		err := fmt.Errorf("expect resource to be %s", StorageClassV1GVR)
+		klog.Error(err)
+		return toV1AdmissionResponse(err)
+	}
+}
+
+func decideV1StorageClass(sc, oldsc *storagev1.StorageClass, isUpdate bool) *v1.AdmissionResponse {
+	reviewResponse := &v1.AdmissionResponse{
+		Allowed: true,
+		Result:  &metav1.Status{},
+	}
+
+	if isUpdate {
+		// TBD
+		return reviewResponse
+	}
+
+	if err := ValidateV1StorageClass(sc); err != nil {
+		reviewResponse.Allowed = false
+		reviewResponse.Result.Message = err.Error()
+	}
+	return reviewResponse
+}
+
+func ValidateV1StorageClass(sc *storagev1.StorageClass) error {
+	if sc == nil {
+		return fmt.Errorf("StorageClass is nil")
+	}
+
+	if sc.Provisioner != FilestoreCSIDriver {
+		return nil
+	}
+
+	params := sc.Parameters
+	if params == nil {
+		return nil
+	}
+
+	if _, ok := params["multishare"]; ok {
+		klog.Infof("Multishare Filestore CSI storage class detected")
+	}
+
+	// TBD: Fill the validation logic here.
+	return nil
+}
+
+func toV1AdmissionResponse(err error) *v1.AdmissionResponse {
+	return &v1.AdmissionResponse{
+		Result: &metav1.Status{
+			Message: err.Error(),
+		},
+	}
+}
+func mutateStorageClass(ar v1.AdmissionReview) *v1.AdmissionResponse {
+	klog.Info("mutating storageClass")
+	reviewResponse := &v1.AdmissionResponse{
+		Allowed: true,
+		Result:  &metav1.Status{},
+	}
+
+	if !(ar.Request.Operation == v1.Update || ar.Request.Operation == v1.Create) {
+		return reviewResponse
+	}
+
+	raw := ar.Request.Object.Raw
+	oldRaw := ar.Request.OldObject.Raw
+
+	deserializer := codecs.UniversalDeserializer()
+	switch ar.Request.Resource {
+	case StorageClassV1GVR:
+		sc := &storagev1.StorageClass{}
+		if _, _, err := deserializer.Decode(raw, nil, sc); err != nil {
+			klog.Error(err)
+			return toV1AdmissionResponse(err)
+		}
+		oldSc := &storagev1.StorageClass{}
+		if _, _, err := deserializer.Decode(oldRaw, nil, oldSc); err != nil {
+			klog.Error(err)
+			return toV1AdmissionResponse(err)
+		}
+		klog.Infof("check patch for storageClass %s", sc.Name)
+		return applyV1StorageClassPatch(sc)
+	default:
+		err := fmt.Errorf("expect resource to be %s", StorageClassV1GVR)
+		klog.Error(err)
+		return toV1AdmissionResponse(err)
+	}
+}
+
+func applyV1StorageClassPatch(sc *storagev1.StorageClass) *v1.AdmissionResponse {
```

Comment:
do we not have to call this in `admitStorageClass` as well?

Reply:
No. See [this](https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/#what-are-admission-webhooks) ``` Mutating admission webhooks are invoked first, and can modify objects sent to the API server to enforce custom defaults. After all object modifications are complete, and after the incoming object is validated by the API server, validating admission webhooks are invoked and can reject requests to enforce custom policies.```